alibi/explainers/similarity/backends/tensorflow/base.py

```diff
@@ -0,0 +1,97 @@
+"""Tensorflow backend for similarity explainers.
+
+Methods unique to the Tensorflow backend are defined here. The interface this class defines syncs with the torch backend
+in order to ensure that the similarity methods only require to match this interface.
+"""
+
+import random
+import os
+from typing import Callable
+
+import numpy as np
+import tensorflow as tf
+import tensorflow.keras as keras
+
+
+class _TensorFlowBackend:
+    device = None
+
+    @staticmethod
+    def get_grads(
+            model: keras.Model,
+            x: tf.Tensor,
+            y: tf.Tensor,
+            loss_fn: Callable[[tf.Tensor, tf.Tensor], tf.Tensor],
+    ) -> np.ndarray:
+        """
+        Computes the gradients of the loss function with respect to the model's parameters for a single training and
+        target pair.
+
+        Parameters:
+        -----------
+        model:
+            The model to compute gradients for.
+        x:
+            The input data point.
+        y:
+            The target data point.
+        loss_fn:
+            The loss function to use.
+
+        Returns:
+        --------
+        grads: np.ndarray
+            The gradients of the loss function with respect to the model's parameters. This is returned as a flattened \
+            array.
+        """
+
+        with tf.device(_TensorFlowBackend.device):
+            with tf.GradientTape() as tape:
+                output = model(x, training=False)
+                loss = loss_fn(y, output)
+
+            # compute gradients of the loss w.r.t the weights
+            grad_x_train = tape.gradient(loss, model.trainable_weights)
+            grad_x_train = np.concatenate([w.numpy().reshape(-1) for w in grad_x_train])
+        return grad_x_train
+
+    @staticmethod
+    def to_tensor(x: np.ndarray, **kwargs) -> tf.Tensor:
+        """Converts a `numpy` array to a `tensorflow` tensor."""
+        return tf.convert_to_tensor(x)
+
+    @staticmethod
+    def set_device(device: str = 'cpu:0') -> None:
```

Comment:
I think with Tensorflow we need to be a little bit careful here. Whilst for `torch` the user must handle the devices for their model parameters and other tensors manually, with `tensorflow` if a GPU is present usually that is taken to be the default device automatically. I'm wondering if by setting a CPU as default here we can run into some unintended consequences if a GPU is present and the model is already on the GPU?

Reply:
yeah, I think you are correct.

I've set it so that if `_TensorflowBackend.device` is `None` which is the default, then the `with tf.device(_TensorflowBackend.device):` should mean tensorflow will default to using its [default device](https://stackoverflow.com/a/41525764). 

I've also made the backend argument explicit rather than a `kwarg` so it's clearer documented that the user can set the device themselves.